### PR TITLE
Fix EsAbortPolicy to conform to API

### LINKED ISSUE #19508: EsAbortPolicy violates the RejectedExecutionHandler API
The interface for `RejectedExecutionHandler` says:

``` java
public interface RejectedExecutionHandler {
    /**
     * Method that may be invoked by a {@link ThreadPoolExecutor} when
     * {@link ThreadPoolExecutor#execute execute} cannot accept a
     * task.  This may occur when no more threads or queue slots are
     * available because their bounds would be exceeded, or upon
     * shutdown of the Executor.
     *
     * <p>In the absence of other alternatives, the method may throw
     * an unchecked {@link RejectedExecutionException}, which will be
     * propagated to the caller of {@code execute}.
     *
     * @param r the runnable task requested to be executed
     * @param executor the executor attempting to execute this task
     * @throws RejectedExecutionException if there is no remedy
     */
    void rejectedExecution(Runnable r, ThreadPoolExecutor executor);
}
```

It's the `@throws` clause that is violated. However, `EsAbortPolicy` throws an `EsRejectedExecutionException` which does not inherit from `RejectedExecutionException`. This leads to situations that commits like 770186f6cfd1a843f6cfca3df791093459b47f4b need to remedy. This should be cleaned up, but it's probably not a small task.

### PR DESCRIPTION
**Pull request by jasontedor**
The rejected execution handler API says that rejectedExecution(Runnable, ThreadPoolExecutor) throws a RejectedExecutionException if the task must be rejected due to capacity on the executor. We do throw something that smells like a RejectedExecutionException (it is named EsRejectedExecutionException) yet we violate the API because EsRejectedExecutionException is not a RejectedExecutionException. This has caused problems before where we try to catch RejectedExecution when invoking rejectedExecution but this causes EsRejectedExecutionException to go uncaught. This commit addresses this by modifying EsRejectedExecutionException to extend RejectedExecutionException. Addtionally, we fix an issue with how causes are unwrapped. Without this change, since EsRejectedExecutionException is no longer an ElasticsearchException, the root cause of a rejection would be reported as a remote_transport_exception instead of as an es_rejected_execution_exception.

Closes #19508

### CODE REVIEW

**Comment by jasontedor** [comment]
retest this please

**Review by bleskes — APPROVED** [review]
LGTM, though I presume the BWC bridge is coming in the 6.3 port?

**Comment by jasontedor** [comment]
> LGTM, though I presume the BWC bridge is coming in the 6.3 port?

That was the plan but after I started preparing that I realized it is a little hairy. I pushed it here and will change the version after I backport. Would you please take a look at the bridge that I pushed here?

**Comment by jasontedor** [comment]
> Addtionally, we fix an issue with how causes are unwrapped.

I think that this change should be made in a separate change so I have pulled it out of this one.

**Comment by jasontedor** [comment]
@bleskes Can you take another look?

File: server/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java [hunk 1]
```diff
@@ -852,8 +853,28 @@ public void writeException(Throwable throwable) throws IOException {
                 writeCause = false;
             } else if (throwable instanceof IOException) {
                 writeVInt(17);
+            } else if (throwable instanceof EsRejectedExecutionException) {
+                // TODO: remove the if branch when master is bumped to 8.0.0
+                assert Version.CURRENT.major < 8;
+                if (version.before(Version.V_7_0_0_alpha1)) {
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java]: It seems like you're not planning to back port this any more. Can you push a comment in the 6.x branch to note that if you change the wire level there, you need to look at 7.x code too? I don't think we will so a comment seems enough (a test is tricky)

**Review by bleskes — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java (hunk 1 above)
Comment (jasontedor) [reply server/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java]: I am planning to backport. I do it this way so BWC tests can pass on this branch. Then I backport and change the version. Then I remove this code from master.

A test is very tricky indeed. I thought about a qa test in 6.x only (with a bulk thread pool with queue 0 to force a rejection). I am not sure it’s worth it. I did test it manually like this.

Comment (bleskes) [reply server/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java]: Agreed that a test is an overkill. At least any test I can think of.

**Comment by jasontedor** [comment]
6.x: d72757fa18d0d217829e94c5d5ff43583ac8cc27
Removing BWC layer from master: 1f1a4d17b4b06fcf3548e5d49e5262725ea1307d

**Pull request merged**